Spanner/src/SpannerClient.php

```diff
@@ -126,6 +126,8 @@ class SpannerClient
      *     @type bool $returnInt64AsObject If true, 64 bit integers will be
      *           returned as a {@see Google\Cloud\Core\Int64} object for 32 bit
      *           platform compatibility. **Defaults to** false.
+     *     @type bool $useGapicBackoffs If true, GAPIC backoff strategy
```

Comment:
WYDT about using the name `useLegacyBackoffs` instead? The term GAPIC isn't really widely known/used outside of our internal use cases. We could default it to true - and when false then employ the GAPIC retry strategy.

It would also be nice to add a note outlining specifically that the flag will enable varied retries per call as opposed to a blanket strategy across all.

Reply:
I see the point, but I'm not sure about the flag inversion. Using GAPIC strategies does not hide or alter legacy backoffs, it just adds a new layer. Flag off = legacy, flag on = extra features - that's the common convention.

If the term "GAPIC" is too cryptic, we can use something like `useDiscreteBackoffs`.

A note would be helpful, but it's hard to make a note both brief and informative. I think of something like that:
`True: when the server returns an error, backoff strategy based on called method name is used before the request is considered failed. False: fail immediately. **Defaults to** false.`

Reply:
> Using GAPIC strategies does not hide or alter legacy backoffs, it just adds a new layer.

I would have thought we would disable the existing (legacy) retry mechanisms, otherwise we could end up retrying the retries from the GAPIC layer - I don't think we'd want that :) (but am open to hearing the argument otherwise).

Reply:
It's not that bad. A user may disable legacy backoffs setting `retries` option to `0`. So, a user may choose which strategy to use, or to use none, or to use both (don't know why would someone need that - but what if?). We could disable legacy backoffs automatically when GAPIC strategy is chosen, but to me this behaviour is too complicated and hard to explain.

Reply:
> We could disable legacy backoffs automatically when GAPIC strategy is chosen, but to me this behaviour is too complicated and hard to explain.

This is the correct approach to me. I don't understand why this would be hard to document or explain. What seems more difficult for users to understand is that the retries they expect to work from the GAPIC layer actually get repeated.

> So, those mechanisms are not interchangeable.

They don't have to be. That is why we keep the non-breaking retry mechanism as the default and allow the breaking one to be opt-in.

Again, I think I would prefer to fix this cleanly in a breaking change but this is something we wanted to do now so we need to make the correct accommodations to make it clear and easy to use for folks.

Reply:
All right, making different strategies mutually exclusive is correct.

In this case wouldn't be better to use an enumeration instead of a flag? So that selecting a strategy would look like
```php
$options = [
    'backoff' => SpannerClient::BACKOFF_DISCRETE,
];
$spanner = new SpannerClient($options);
```
It's a bit verbose, but this way it is more clear that different backoff mechanisms won't stack.

Reply:
An enum seems to make the most sense if we expect there to be >2 retry strategies, but I think we're just evaluating 2 with the plan to move to 1 with a breaking change in the future. 

> this way it is more clear that different backoff mechanisms won't stack.

I'm not saying it isn't more clear, I'm just having trouble seeing it right now - why does the enum make things more clear than some documentation?

Reply:
All right. If we won't need no more strategies absolutely never - a flag will do.

Reply:
Thanks @ava12, I really like your suggestion of `useDiscreteBackoffs` earlier.